File: tools/convert-png-to-webp.mjs
import { readdir, readFile, rm, stat, writeFile } from "node:fs/promises";
import path from "node:path";
import sharp from "sharp";

const rootDir = process.cwd();
const sourceRoot = path.resolve(rootDir, "source");
const textRoots = [
  sourceRoot,
  path.resolve(rootDir, "_config.next.yml"),
];
const textExtensions = new Set([".css", ".html", ".js", ".json", ".yml", ".yaml"]);

async function walk(dir, predicate, files = []) {
  const entries = await readdir(dir, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      await walk(fullPath, predicate, files);
      continue;
    }

    if (entry.isFile() && predicate(fullPath)) {
      files.push(fullPath);
    }
  }

  return files;
}

function toPosixPath(filePath) {
  return filePath.split(path.sep).join("/");
}

function toSitePath(filePath) {
  return `/${toPosixPath(path.relative(sourceRoot, filePath))}`;
}

function selectWebpOptions(filePath, metadata, bytes) {
  const normalizedPath = toPosixPath(filePath);
  const maxEdge = Math.max(metadata.width ?? 0, metadata.height ?? 0);

  if (metadata.hasAlpha) {
    return { lossless: true, effort: 6 };
  }

  if (normalizedPath.includes("/assets/images/icons/")) {
    return { lossless: true, effort: 6 };
  }

  if (bytes <= 128 * 1024 && maxEdge <= 512) {
    return { lossless: true, effort: 6 };
  }

  return {
    quality: 95,
    alphaQuality: 100,
    effort: 6,
    smartSubsample: true,
  };
}

function replaceAll(content, search, replacement) {
  return content.split(search).join(replacement);
}

async function listTextFiles() {
  const files = [];

  for (const target of textRoots) {
    const targetStat = await stat(target);

    if (targetStat.isDirectory()) {
      const nested = await walk(target, (filePath) => textExtensions.has(path.extname(filePath).toLowerCase()));
      files.push(...nested);
      continue;
    }

    files.push(target);
  }

  return files;
}

const pngFiles = await walk(sourceRoot, (filePath) => path.extname(filePath).toLowerCase() === ".png");
const replacements = [];
let totalOriginalBytes = 0;
let totalWebpBytes = 0;

for (const pngFile of pngFiles) {
  const image = sharp(pngFile, { animated: true });
  const metadata = await image.metadata();
  const originalBytes = (await stat(pngFile)).size;
  const webpFile = pngFile.replace(/\.png$/i, ".webp");
  const webpOptions = selectWebpOptions(pngFile, metadata, originalBytes);

  await image.webp(webpOptions).toFile(webpFile);

  const webpBytes = (await stat(webpFile)).size;
  totalOriginalBytes += originalBytes;
  totalWebpBytes += webpBytes;

  replacements.push({
    fromSitePath: toSitePath(pngFile),
    toSitePath: toSitePath(webpFile),
    fromSourcePath: toPosixPath(path.relative(sourceRoot, pngFile)),
    toSourcePath: toPosixPath(path.relative(sourceRoot, webpFile)),
    fromRelativePath: toPosixPath(path.relative(rootDir, pngFile)),
    toRelativePath: toPosixPath(path.relative(rootDir, webpFile)),
  });
}

const textFiles = await listTextFiles();
let changedTextFiles = 0;

for (const textFile of textFiles) {
  const original = await readFile(textFile, "utf8");
  let next = original;

  for (const replacement of replacements) {
    next = replaceAll(next, replacement.fromSitePath, replacement.toSitePath);
    next = replaceAll(next, replacement.fromSourcePath, replacement.toSourcePath);
    next = replaceAll(next, replacement.fromRelativePath, replacement.toRelativePath);
  }

  if (next !== original) {
    await writeFile(textFile, next, "utf8");
    changedTextFiles += 1;
  }
}

for (const pngFile of pngFiles) {
  await rm(pngFile);
}

const savedBytes = totalOriginalBytes - totalWebpBytes;
console.log(`Converted ${pngFiles.length} PNG files to WebP.`);
console.log(`Updated ${changedTextFiles} text files with new image paths.`);
console.log(`Original size: ${(totalOriginalBytes / (1024 * 1024)).toFixed(2)} MB`);
console.log(`WebP size: ${(totalWebpBytes / (1024 * 1024)).toFixed(2)} MB`);
console.log(`Saved: ${(savedBytes / (1024 * 1024)).toFixed(2)} MB`);

File: tools/convert-png-to-webp.py
from __future__ import annotations

from pathlib import Path
from typing import Iterable

from PIL import Image


ROOT_DIR = Path(__file__).resolve().parent.parent
SOURCE_ROOT = ROOT_DIR / "source"
TEXT_ROOTS = [
    SOURCE_ROOT,
    ROOT_DIR / "_config.next.yml",
]
TEXT_EXTENSIONS = {".css", ".html", ".js", ".json", ".yml", ".yaml"}


def walk_png_files(root: Path) -> Iterable[Path]:
    for path in root.rglob("*.png"):
        if path.is_file():
            yield path


def walk_text_files() -> Iterable[Path]:
    for target in TEXT_ROOTS:
        if target.is_dir():
            for path in target.rglob("*"):
                if path.is_file() and path.suffix.lower() in TEXT_EXTENSIONS:
                    yield path
            continue

        if target.is_file():
            yield target


def to_posix(path: Path) -> str:
    return path.as_posix()


def to_site_path(path: Path) -> str:
    return "/" + to_posix(path.relative_to(SOURCE_ROOT))


def select_webp_options(path: Path, image: Image.Image, size_bytes: int) -> dict:
    normalized_path = to_posix(path)
    width, height = image.size
    max_edge = max(width, height)
    has_alpha = image.mode in ("RGBA", "LA") or ("transparency" in image.info)

    if has_alpha:
        return {"lossless": True, "method": 6}

    if "/assets/images/icons/" in normalized_path:
        return {"lossless": True, "method": 6}

    if size_bytes <= 128 * 1024 and max_edge <= 512:
        return {"lossless": True, "method": 6}

    return {"quality": 95, "method": 6}


def replace_all(content: str, search: str, replacement: str) -> str:
    return content.replace(search, replacement)


def main() -> None:
    png_files = list(walk_png_files(SOURCE_ROOT))
    replacements: list[dict[str, str]] = []
    total_original_bytes = 0
    total_webp_bytes = 0

    for png_file in png_files:
        original_bytes = png_file.stat().st_size
        webp_file = png_file.with_suffix(".webp")

        with Image.open(png_file) as image:
            converted = image
            if image.mode not in ("RGB", "RGBA", "LA"):
                converted = image.convert("RGBA" if "transparency" in image.info else "RGB")

            save_options = select_webp_options(png_file, converted, original_bytes)
            converted.save(webp_file, format="WEBP", **save_options)

        webp_bytes = webp_file.stat().st_size
        total_original_bytes += original_bytes
        total_webp_bytes += webp_bytes

        replacements.append(
            {
                "from_site_path": to_site_path(png_file),
                "to_site_path": to_site_path(webp_file),
                "from_source_path": to_posix(png_file.relative_to(SOURCE_ROOT)),
                "to_source_path": to_posix(webp_file.relative_to(SOURCE_ROOT)),
                "from_relative_path": to_posix(png_file.relative_to(ROOT_DIR)),
                "to_relative_path": to_posix(webp_file.relative_to(ROOT_DIR)),
            }
        )

    changed_text_files = 0

    for text_file in walk_text_files():
        original = text_file.read_text(encoding="utf-8")
        updated = original

        for replacement in replacements:
            updated = replace_all(updated, replacement["from_site_path"], replacement["to_site_path"])
            updated = replace_all(updated, replacement["from_source_path"], replacement["to_source_path"])
            updated = replace_all(updated, replacement["from_relative_path"], replacement["to_relative_path"])

        if updated != original:
            text_file.write_text(updated, encoding="utf-8")
            changed_text_files += 1

    for png_file in png_files:
        png_file.unlink()

    saved_bytes = total_original_bytes - total_webp_bytes
    print(f"Converted {len(png_files)} PNG files to WebP.")
    print(f"Updated {changed_text_files} text files with new image paths.")
    print(f"Original size: {total_original_bytes / (1024 * 1024):.2f} MB")
    print(f"WebP size: {total_webp_bytes / (1024 * 1024):.2f} MB")
    print(f"Saved: {saved_bytes / (1024 * 1024):.2f} MB")


if __name__ == "__main__":
    main()
